corehq/apps/app_manager/fields.py

```diff
@@ -116,3 +139,373 @@ def get_app_sources(domain):
+        return chosen_by_choice
+
+    @staticmethod
+    def _get_item_name(item, has_app, app_lang, default_name):
+        item_name = None
+        if has_app and item is not None:
+            item_name = item['name'].get(app_lang) or item['name'].get('en')
+        return item_name or default_name
+
+    def _get_modules_and_forms(self, as_dict=True):
+        modules_by_app = collections.defaultdict(list)
+        forms_by_app_by_module = {}
+        for form in self._all_forms:
+            has_app = form.get('has_app', False)
+            app_lang = form['app']['langs'][0] if 'langs' in form['app'] else 'en'
+            app_id = form['app']['id'] if has_app else self.UNKNOWN_SOURCE
+            module = form.get('module')
+            module_id = (module['id'] if has_app and module is not None
+                         else self.UNKNOWN_SOURCE)
+            module_name = self._get_item_name(
+                module, has_app, app_lang, _("Unknown Module")
+            )
+            form_xmlns = form['xmlns']
+            form_name = form_xmlns
+            if not form.get('show_xmlns', False):
+                form_name = self._get_item_name(
+                    form.get('form'), has_app, app_lang,
+                    "{} (potential matches)".format(form_xmlns)
+                )
+            module_choice = RMIDataChoice(
+                module_id,
+                module_name,
+                form
+            )
+            form_choice = RMIDataChoice(
+                form_xmlns,
+                form_name,
+                form
+            )
+            if as_dict:
+                form_choice = form_choice._asdict()
+
+            if app_id not in forms_by_app_by_module:
+                forms_by_app_by_module[app_id] = collections.defaultdict(list)
+            modules_by_app[app_id].append(module_choice)
+            forms_by_app_by_module[app_id][module_id].append(form_choice)
+
+        modules_by_app = self._get_unique_choices(modules_by_app)
+        if as_dict:
+            modules_by_app = self._map_chosen_by_choice_as_dict(modules_by_app)
+        return modules_by_app, forms_by_app_by_module
+
+    def get_form_rmi_response(self):
+        """Use this to generate the response that initializes the form
+        controlled by hq.app_data_drilldown.ng.js if you are drilling down
+        to an XForm + app_id pair"""
+        modules_by_app, forms_by_app_by_module = self._get_modules_and_forms(self.as_dict)
+        response = AppFormRMIResponse(
+            app_types=self._get_app_type_choices(self.as_dict),
+            apps_by_type=self._get_applications_by_type(self.as_dict),
+            modules_by_app=modules_by_app,
+            forms_by_app_by_module=forms_by_app_by_module,
+            placeholders=self.form_placeholders,
+        )
+        if self.as_dict:
+            response = response._asdict()
+        return response
+
+    def _get_cases_for_apps(self, apps_by_type, as_dict=True):
```

Comment:
minor, but i'm generally not a huge fan of methods that take in a parameter that changes their return types. a pattern i like if the code reuse is high enough to justify that is to make two functions that are intended to be called with really clear names, and then only allow those functions to call the function with the param.

so like
```python
def _get_cases_for_apps_as_dicts(self, apps_by_type):
  return _get_cases_for_apps(self, apps_by_type, as_dict=True)

def _get_cases_for_apps_as_rmi_choices(self, apps_by_type):
  return _get_cases_for_apps_as_dicts(self, apps_by_type, as_dict=False)
```